### PR TITLE
Fix tick_calendar and passing datetime as parameter (!=timestamps)

### PR DESCRIPTION
**Pull request by achoum**

### CODE REVIEW

**Comment by javiber** [comment]
It's a different bug, @achoum fixed something related to DST and mine was related to some missing ticks at the beginning of the span.

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils_test.cc [hunk 1]
```diff
@@ -0,0 +1,19 @@
+
+#include "temporian/implementation/numpy_cc/operators/tick_calendar_utils.h"
+
+#include "gtest/gtest.h"
+
+TEST(IsLeapYear, Base) {
+  EXPECT_TRUE(IsLeapYear(1972));
```
Comment (rstz) [inline temporian/implementation/numpy_cc/operators/tick_calendar_utils_test.cc:7]: Add 2000 and 1900 as tests to increase coverage

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils_test.cc [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+  EXPECT_FALSE(IsLeapYear(1970));
+}
+
+TEST(UTCMkTime, Base) {
+  // Tested with www.epochconverter.com
+  EXPECT_EQ(UTCMkTime(1970, 1, 1, 0, 0, 0).value().seconds_since_epoch, 0);
```
Comment (rstz) [inline temporian/implementation/numpy_cc/operators/tick_calendar_utils_test.cc:13]: Can you add a few more tests, e.g. a test before 1970 and a test after the year 3000? And one in Summer time and one in winter time

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.h [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+#include <optional>
+
+struct MyTime {
+  int64_t seconds_since_epoch;
+  int wday;
+};
+
+bool IsLeapYear(int year);
+
+std::optional<MyTime> UTCMkTime(const int year, const int month, const int day,
```
Comment (rstz) [inline temporian/implementation/numpy_cc/operators/tick_calendar_utils.h:11]: Add documentation here?

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc [hunk 4]
```diff
@@ -0,0 +1,54 @@
+#include "temporian/implementation/numpy_cc/operators/tick_calendar_utils.h"
+
+#include <cstdint>
+#include <optional>
+
+// Number of days in each month (non-leap year)
+constexpr int daysPerMonth[12] = {31, 28, 31, 30, 31, 30,
+                                  31, 31, 30, 31, 30, 31};
+
+bool IsLeapYear(int year) {
```
Comment (rstz) [inline temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:10]: const int year

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils_test.cc (hunk 2 above)
Comment (rstz) [reply temporian/implementation/numpy_cc/operators/tick_calendar_utils_test.cc:13]: And one that's on February 29

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc [hunk 5]
```diff
@@ -0,0 +1,54 @@
...
+  return ((year % 4) == 0) && ((year % 100) != 0 || (year % 400) == 0);
+}
+
+std::optional<MyTime> UTCMkTime(const int year, const int month, const int day,
+                                const int hour, const int minute,
+                                const int second) {
+  if (month < 1 || month > 12 || day > daysPerMonth[month - 1]) {
```
Comment (rstz) [inline temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:17]: IIUC, UTCMkTime(2000, 2, 29, 0, 0, 1) will return {}, which looks like a bug.

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc [hunk 6]
```diff
@@ -0,0 +1,54 @@
...
+    // Invalid date.
+    return {};
+  }
+
+  const int64_t seconds_per_days = 24 * 60 * 60;
```
Comment (rstz) [inline temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:22]: Nit: seconds_per_days --> seconds_per_day
Second nit: constexpr

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc [hunk 7]
```diff
@@ -0,0 +1,54 @@
...
+    seconds_since_epoch += seconds_per_days;
+  }
+
+  seconds_since_epoch += (day - 1) * seconds_per_days;
+  seconds_since_epoch += hour * 60 * 60;
+  seconds_since_epoch += minute * 60;
+  seconds_since_epoch += second;
+
+  const auto days_since_epoch = seconds_since_epoch / seconds_per_days;
+  // Note: 1970-1-1 was a thursday.
```
Comment (rstz) [inline temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:50]: Nit: thursday --> Thursday

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc (hunk 5 above)
Comment (rstz) [inline temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:17]: I think more checks are useful here, e.g. what if days is negative or year < 1970 ? Not sure what's expected to get returned by this function

**Review by rstz — CHANGES_REQUESTED** [review]

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils_test.cc (hunk 2 above)
Comment (achoum) [reply temporian/implementation/numpy_cc/operators/tick_calendar_utils_test.cc:13]: *Added all requested test and a few others. Added test for invalid dates.

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.h (hunk 3 above)
Comment (achoum) [reply temporian/implementation/numpy_cc/operators/tick_calendar_utils.h:11]: *Added comments

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc (hunk 4 above)
Comment (achoum) [reply temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:10]: Good catch.
*Done

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc (hunk 5 above)
Comment (achoum) [reply temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:17]: *Solved + added more tests to check more similar cases.

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc (hunk 6 above)
Comment (achoum) [reply temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:22]: *Replace + constexpr

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc (hunk 7 above)
Comment (achoum) [reply temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:50]: *Done

File: temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc (hunk 5 above)
Comment (achoum) [reply temporian/implementation/numpy_cc/operators/tick_calendar_utils.cc:17]: Year before 1970 are good here (but there were an issue with leap years in this case.)
* Fix leap year bug before 1970.
*Added a test for leap year.
*Added more tests to show it works as expected.

**Comment by achoum** [comment]
Thanks for the initial review.

**Review by rstz — APPROVED** [review]
Thank you for the changes, LGTM

**Comment by achoum** [comment]
Thanks

**Pull request merged**